### PR TITLE
pid1: add a new method of rebooting: userspace only under the name "soft-reboot"

### PR DESCRIPTION
**Pull request by poettering**
This adds a new mechanism for rebooting, a form of "userspace reboot"
    hereby dubbed "soft-reboot". It will stop all services as in a usual
    shutdown, possibly transition into a new root fs and then issue a fresh
    initial transaction. The kernel is not replaced.
    
    File descriptors can be passed over, thus opening the door for leaving
    certain resources around between such reboots.
    
    Usecase: this is an extremely quick way to reset userspace fully when
    updating image based systems, without going through a full
    hardware/firmware/boot loader/kernel reset. It minimizes "grayout time"
    for OS updates. (In particular when combined with kernel live patching)

### CODE REVIEW

**Comment by poettering** [comment]
this is not ready for merging yet. lacks all docs, and tests.

But is pretty comprehensive otherwise. Can do a supercharged reboot in an nspawn container and in qemu in no time.

**Comment by bluca** [comment]
This is great, and should open the door to easier testing as well here, as we can now do reboots without actually terminating qemu/nspawn and thus add special logic to the test harness.

Bikeshedding: I'd really call this just "userspace reboot", "renew" is really confusing if you don't already know what it does

**Comment by poettering** [comment]
"systemctl userspace-reboot" is impossible to type

**Comment by bluca** [comment]
That's why god invented bash completion!

**Comment by bluca** [comment]
As a shorthand, we could also have a "systemctl uexec" alias - would pair well with kexec

**Comment by srd424** [comment]
> "systemctl userspace-reboot" is impossible to type

How about "reinit"?

**Comment by orowith2os** [comment]
`reinit` sounds too much like a reboot to me, from a third party (end user) pov.

I prefer `userspace-reboot`, it feels like it represents this feature best. It's easy to remember. I can't think of any alternative names, though...

**Comment by dtardon** [comment]
Maybe `system-reexec` (analogy to `daemon-reexec`)? Or just `reexec`?

Btw, `launchd` apparently supports this already; it's invoked as `launchctl reboot userspace`.

**Comment by bluca** [comment]
the main goal is to quickly transition to a new userspace snapshot, not just to re-execute the current one

**Comment by mrc0mmand** [comment]
You'll need to add

```diff
diff --git a/test/units/testsuite-21.sh b/test/units/testsuite-21.sh
index 36f647ca5f..a0df607377 100755
--- a/test/units/testsuite-21.sh
+++ b/test/units/testsuite-21.sh
@@ -28,6 +28,7 @@ systemctl log-level info
 # FIXME: systemd-run doesn't play well with daemon-reexec
 # See: https://github.com/systemd/systemd/issues/27204
 sed -i '/\[org.freedesktop.systemd1\]/aorg.freedesktop.systemd1.Manager:Reexecute FIXME' /etc/dfuzzer.conf
+sed -i '/\[org.freedesktop.systemd1\]/aorg.freedesktop.systemd1.Manager:Renew destructive' /etc/dfuzzer.conf
 
 # TODO
 #   * check for possibly newly introduced buses?
```

so dfuzzer won't keep triggering "renew" when fuzzing. Once this is merged (and the name of the method is settled) I'll update dfuzzer and drop the sed.

**Comment by poettering** [comment]
> Btw, `launchd` apparently supports this already; it's invoked as `launchctl reboot userspace`.

btw, anyone has any idea where to find the current sources for launchd/launchtl? did apple take that closed source? only can find a verson from 7 years ago...

**Comment by poettering** [comment]
> ```diff
> ```diff
> 
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> so dfuzzer doesn't keep triggering "renew" when fuzzing. Once this is merged (and the name of the method is settled) I'll update dfuzzer and drop the sed.
> ```

apparently people don't like the name "renew". so we are going to change this before merging anyway, i guess

**Comment by Winterhuman** [comment]
@poettering Out of curiosity, how would switching to a different root work in this case? As far as I understand, `switch_root` actually "removes" all the files and dirs under `/` and populates it with the new root's content, as used in initrd's on tmpfs transitioning into the real system, but here, that would remove the old root's actual content and wouldn't work.

How does a non-destructive `switch_root` work? (unless this isn't the right place to ask)

**Comment by topimiettinen** [comment]
More bikeshedding suggestions:
- Also analogously to `systemctl daemon-reexec`: `systemctl userspace-reexec` or less Yoda: `systemctl reexec-userspace`?
- analogously to `systemctl restart xyz.service`: `systemctl userspace-restart` / `systemctl restart-userspace`
- `systemctl userspace-reboot` / `systemctl reboot-userspace-only`/ `systemctl reboot-userspace`

Also the existing commands could have new, complementing aliases for symmetry:
-  `systemctl reboot` = `systemctl system-reboot` / `systemctl reboot-system` / `systemctl system-restart` / `systemctl restart-system`
-  `systemctl kexec` = `systemctl kernel-reboot` / `systemctl reboot-kernel` / `systemctl kernel-restart` / `systemctl restart-kernel`
-  `systemctl poweroff` = `systemctl system-poweroff` / `systemctl poweroff-system` (maybe not, what's the userspace equivalent to poweroff?)

**Comment by jevinskie** [comment]
> > Btw, `launchd` apparently supports this already; it's invoked as `launchctl reboot userspace`.
> 
> btw, anyone has any idea where to find the current sources for launchd/launchtl? did apple take that closed source? only can find a verson from 7 years ago...

Sadly they stopped releasing the source years ago.

File: src/core/manager.c
Comment (bew) [inline src/core/manager.c]: note: can remove this comment, there would be no space anymore here
(the next one is `SIGRTMIN+18` which is used)

**Comment by int-72h** [comment]
>     * `systemctl poweroff` = `systemctl system-poweroff` / `systemctl poweroff-system`  (maybe not, what's the userspace equivalent to poweroff?)

One could argue it'd be dropping to emergency mode.

**Comment by cyqsimon** [comment]
I must say I also dislike the name "renew", in particular because the word is already commonly used in the context of a subscription, service, or certificate renewal. However in this context the word "renew" has a meaning more closely resembling "continuation" or "extension", whereas in our context the intended meaning is "restart" (albeit just a different kind). In short, it certainly would be very confusing for many users.

Personally I much prefer the previously suggested "reinit". And also, have you considered "refresh"? It's already commonly used in the context of refreshing a webpage, where the meaning is actually ***very*** similar to our use case - restarting the userspace (analogous to the webpage) without touching the kernel (analogous to the browser).

**Comment by vaparetia** [comment]
I have a suggestion, what about the naming the command like this: "systemctl reboot --userspace" adding a flag to an already existing command since that's what you are thinking when you execute it. 

Anyway, just a suggestion. This feature seems really cool.

File: src/core/main.c
Comment (yuwata) [inline src/core/main.c]: missing `return r;` ?

File: src/shutdown/shutdown.c
Comment (yuwata) [inline src/shutdown/shutdown.c]: 'bew' -> 'new'

**Comment by Doomsdayrs** [comment]
> I have a suggestion, what about the naming the command like this: "systemctl reboot --userspace" adding a flag to an already existing command since that's what you are thinking when you execute it.
> 
> Anyway, just a suggestion. This feature seems really cool.

I was thinking the same, RFC on the feasibility of such

**Comment by webbb82** [comment]
Here is the new name.  **_"USReboot_** **U**_ser_ **S**_pace_ **R**_eboot"_ or "**SUR** <u>**S**</u>ystemctl <u>**U**</u>ser **<u>**R**</u>**eboot"  simple, clean, rolls of the tongue nicely.

**Comment by poettering** [comment]
Thank you all for the input. The cabal has spoken, and voted for the name "soft-reboot", hence I renamed it now accordingly. (My own preference for "renew" didn't fare particularly well, sniff). Hence, let's end the bikeshedding...

**Comment by poettering** [comment]
> @poettering Out of curiosity, how would switching to a different root work in this case? In other words, how does a non-destructive `switch_root` work?

The destructive switch-root we do at boot is destructive because we basically rm-rf the contents of the original initrd tmpfs after switching root. For this "soft-reboot" logic we skip that step, that's all, hence we aren't destructive.

Note that the Linux kernel currently pins the original root fs, no matter what, hence we cannot properly get rid of the initrd tmpfs. All we can do is empty it, as described, and overmount it with a new file system. However, when we transition from that fs to the next (as per soft-reboot) then we can get rid of the old one, simply by unmounting it fully.

**Comment by poettering** [comment]
test added now, too. should be complete now and ready for review. ptal!

**Comment by poettering** [comment]
> > Note that the Linux kernel currently pins the original root fs, no matter what
> 
> Do you have any resources on the status of changing this upstream? 

I know that at least one of the Linux kernel VFS maintainers is aware of the issue. I wouldn't hold my breath though.

> Would an image-based system like ostree help here?

ostree is a userspace concept. It cannot avoid this kernel limitation. This needs to be fixed in kernel.

**Comment by erkinalp** [comment]
This should have been called `systemctl --user reboot`.

**Comment by smac89** [comment]
> This should have been called `systemctl --user reboot`.

`--user` commands usually do not require super user permissions, but I have a feeling this feature will require root permissions

**Comment by bluca** [comment]
needs a rebase

**Comment by poettering** [comment]
> This should have been called `systemctl --user reboot`.

Generally, if you specify `--user` on the `systemctl` command line you will talk to the per-user instance of the service manager (as supposed to the per-system service manager, i.e. PID 1). Hence `systemctl --user reboot` as user `lennart`would indicate "Hey, service manager of user `lennart` please reboot yourself!". Which doesn't really make sense, since "reboot" is not a concept defined for users, but only for the system as a whole.

**Comment by erkinalp** [comment]
It would indeed make sense, by forcibly restarting everything run by the user in question.

File: src/core/dbus-manager.c
Comment (bluca) [inline src/core/dbus-manager.c]: include `root` in the log message?

File: man/systemd-soft-reboot.service.xml
Comment (bluca) [inline man/systemd-soft-reboot.service.xml]: maybe plug in portable services and nspawn as solutions that work well with this, and link their docs?

File: man/systemd-soft-reboot.service.xml
Comment (bluca) [inline man/systemd-soft-reboot.service.xml]: "not executed, the executables in"

File: test/units/testsuite-82.sh
Comment (bluca) [inline test/units/testsuite-82.sh:115]: would be good to beef up this test, to add a normal service that is shut down as expected (and verify that it is), and one that survives (and verify that it does)

**Review by bluca — CHANGES_REQUESTED** [review]
I wonder what to do about pcrphase and such - I think we do not want to re-run TPM measurements? To achieve this, and probably to help with other actions that might be needed to run only the "real" boot, what about adding a new ConditionIsSoftReboot= or ConditionBootType= or so, that allows to match and skip? To implement it, it should be doable to pass a variable through via serialization in the manager at shutdown, so that the next iteration knows it's been through a soft reboot

**Comment by orowith2os** [comment]
> It would indeed make sense, by forcibly restarting everything run by the user in question.

You can log out and log back in. That should get you what you want. Userspace reboots should probably be under the normal reboot commands.

**Comment by poettering** [comment]
> I wonder what to do about pcrphase and such - I think we do not want to re-run TPM measurements? 

So that's a major discussion to be had, but I think we should do that separately. That said I nowadays think the only safe and secure way is that we really should run at least the boot phase stuff, so that the n-th reboot can be securely distinguished from the n-1-th and n+1-th boot. But we need to start maintaining a proper log of measurements sooner rather than later, so that people can verify this. Soft reboots *must* appear as a new series of boot phases in the measurement logs I think.

Note that the keys for FDE and such are already unlocked on a soft reboot, and we don't need to unlock them again, hence the boot phase stuff should work out quite OK still: you'd bind FDE unlocking to the initrd boot phase of the first boot, and then the FDE will work in perpetuity even though we can never access the key in the TPM anymore.

A different story is the measurement of machine-id and the mount uuids into PCR 15. We probably should not repeat that, and we don't really have to if it stays the same and mounted.

> To achieve this, and probably to help with other actions that might be needed to run only the "real" boot, what about adding a new ConditionIsSoftReboot= or ConditionBootType= or so, that allows to match and skip? To implement it, it should be doable to pass a variable through via serialization in the manager at shutdown, so that the next iteration knows it's been through a soft reboot

We can certainly consider that. For the fs uuid measurement an alternative would be to simply leave the service up till the very end, so that on next boot it is still up because serialization, and then we won't rerun it.

**Comment by jluebbe** [comment]
Is there a defined interface for the new userspace to find out how was booted ("normally" or via soft-reboot)? Similarly, where it was booted from, analogous to /proc/cmdline (besides inspecting mountinfo in proc)?

Or would the old userspace prepare that information together with `/run/nextroot` somewhere readable by the new userspace?

I've not seen it in the code, but perhaps I've missed something.

**Comment by brotaxt** [comment]
> ```
> This adds a new mechanism for rebooting, a form of "userspace reboot"
> hereby dubbed "soft-reboot". It will stop all services as in a usual
> shutdown, possibly transition into a new root fs and then issue a fresh
> initial transaction. The kernel is not replaced.
> 
> File descriptors can be passed over, thus opening the door for leaving
> certain resources around between such reboots.
> 
> Usecase: this is an extremely quick way to reset userspace fully when
> updating image based systems, without going through a full
> hardware/firmware/boot loader/kernel reset. It minimizes "grayout time"
> for OS updates. (In particular when combined with kernel live patching)
> ```

Could you specify what is meant by image based operating system?

**Comment by Winterhuman** [comment]
@brotaxt For example, using two read-only SquashFS partitions (the disk "images"), one with the current root, and the other with the updated root to reboot into

File: TODO
Comment (safinaskar) [inline TODO]: It seems you meant `switch-root`, not `switch-reboot`

**Comment by poettering** [comment]
> Is there a defined interface for the new userspace to find out how was booted ("normally" or via soft-reboot)? 

No, there currently is not. What's your specific usecase for that? I mean, we could theoretically add a "soft boot counter" or so, which is 0 on the initial boot, and is increased in each subsequent soft reboot, but if we do I'd hear a convincing usecase for that first.

> Similarly, where it was booted from, analogous to /proc/cmdline (besides inspecting mountinfo in proc)?

with the upcomin realease we expose "bootctl -R" which gives you the backing device that is actually in use right now. It will derive the info from the file system that backs the root dir, and trace it through complex storage, if that's possible. What more would you need?

> Or would the old userspace prepare that information together with /run/nextroot somewhere readable by the new userspace?

I am not a fan of multiple sources of truth. The kernel provides enough info, which we expose via "bootctl -R", I am not sure we need additional sources of truth.

Note that root= on the kernel cmdline is an optional concept, if you omit it, systemd-gpt-generator will automatically search for a suitable root fs. Hence, relying on root= doesn't really work anyway, "bootctl -R" is much more reliable there.

File: TODO
Comment (poettering) [reply TODO]: true. will fix.

File: man/systemd-soft-reboot.service.xml
Comment (poettering) [reply man/systemd-soft-reboot.service.xml]: nspawn isn't that great for this, because it runs a supervisor process for which we (currently) have no good concept of updating. portable services otoh do not have that (their supervisor is pid1 after all, and that we do know how to replace). hence added a mention of those, but left nay mention of nspawn out for now.

Comment (bluca) [reply man/systemd-soft-reboot.service.xml]: Yeah, we think about how to do something about nspawn to make it work nicely in this use case, separately of course - maybe implement daemon-reexec for it too?

Comment (poettering) [reply man/systemd-soft-reboot.service.xml]: yeah, reexec would probably make sense

**Comment by bluca** [comment]
> > I wonder what to do about pcrphase and such - I think we do not want to re-run TPM measurements?
> 
> So that's a major discussion to be had, but I think we should do that separately. That said I nowadays think the only safe and secure way is that we really should run at least the boot phase stuff, so that the n-th reboot can be securely distinguished from the n-1-th and n+1-th boot. But we need to start maintaining a proper log of measurements sooner rather than later, so that people can verify this. Soft reboots _must_ appear as a new series of boot phases in the measurement logs I think.
> 
> Note that the keys for FDE and such are already unlocked on a soft reboot, and we don't need to unlock them again, hence the boot phase stuff should work out quite OK still: you'd bind FDE unlocking to the initrd boot phase of the first boot, and then the FDE will work in perpetuity even though we can never access the key in the TPM anymore.

This doesn't work for sealed credentials though? These need to be un-sealable at any time when a service starts?

File: man/systemd-soft-reboot.service.xml
Comment (bluca) [inline man/systemd-soft-reboot.service.xml:35]: For mount points this should specify how to ensure they are not automatically cleaned up. By default just mounting something on /run/nextroot is not enough, as it will be automatically umounted during the soft reboot.

File: src/core/main.c
Comment (bluca) [inline src/core/main.c]: this will try to ms_move /run to /run/nextroot/run (for the default case, unless a different nextroot is specified), and the kernel doesn't like that as it gets into a loop:

`May 15 17:42:44 localhost systemd[1]: Failed to mount /run to /run/nextroot/run: Too many levels of symbolic links`

File: man/systemd-soft-reboot.service.xml
Comment (bluca) [reply man/systemd-soft-reboot.service.xml:35]: The suggestion to use `-o x-initrd=mount` works, and should probably be documented here. Also, I think we want to automatically umount /run/nextroot after the switch, so that it's cleaned up and not left laying around.

**Comment by jluebbe** [comment]
My background for these questions are embedded systems which use RAUC for atomic A/B updates. Currently, we do a full reboot into the newly installed system (or a kexec in some cases). If the update doesn't change the kernel binary (which we could detect reliably by comparing kernel slot hashes), we could avoid the reboot and use this mechanism instead, nicely avoiding some downtime.

> > Is there a defined interface for the new userspace to find out how was booted ("normally" or via soft-reboot)?
> 
> No, there currently is not. What's your specific usecase for that? I mean, we could theoretically add a "soft boot counter" or so, which is 0 on the initial boot, and is increased in each subsequent soft reboot, but if we do I'd like to hear a convincing usecase for that first.

To perform an update, RAUC needs to know which A/B slot is currently booted (or "active"). In most cases, the bootloader communicates this information to Linux as cmdline parameters, either `root=/dev/foo` or an explict `rauc.slot=root.a` for cases where no simple rootfs device exists). With this information, RAUC knows can derive which inactive slots can be overwritten safely.

My first idea was that I'd need to know when `/proc/cmdline` corresponds to the current rootfs (on normal boot) or not (soft-reboot). What I actually need is the rootfs info, so the approach you suggest below sounds much better.

> > Similarly, where it was booted from, analogous to /proc/cmdline (besides inspecting mountinfo in proc)?
> 
> with the upcoming release we expose "bootctl -R" which gives you the backing device that is actually in use right now. It will derive the info from the file system that backs the root dir, and trace it through complex storage, if that's possible. What more would you need?

I was not aware of that option. From reading the code, it seems like it should give me the needed information.

Why was it added to `bootctl`, though? That is only available with `ENABLE_BOOTLOADER==1`, which is dependent on EFI and bootloader support in systemd (which we don't use). Nothing in the code for `-R/-RR` seems to depend on EFI and systemd bootloader integration, though.

> > Or would the old userspace prepare that information together with /run/nextroot somewhere readable by the new userspace?
> 
> I am not a fan of multiple sources of truth. The kernel provides enough info, which we expose via "bootctl -R", I am not sure we need additional sources of truth.

Of course. Multiple sources of truth should be avoided.

> Note that root= on the kernel cmdline is an optional concept, if you omit it, systemd-gpt-generator will automatically search for a suitable root fs. Hence, relying on root= doesn't really work anyway, "bootctl -R" is much more reliable there.

For EFI-based systems, I definitely agree. For embedded Linux systems using A/B rootfs, all partitions are usually explicitly set up, so an automatic search would add complexity and can be avoided. With fixed partitions, an explicit cmdline parameter from the bootloader (which did the A/B selection) is straight-forward and easy to review. :)

So, having the `-R/-RR` functionality without pulling in EFI or systemd's bootloader support seems to be the missing piece for using soft-reboot with RAUC (or other embedded A/B updaters). Perhaps there is a better place for it than in `bootctl`? Your commit message in c56be2c294f55545ea34417a2ec4f6ad2cd1df6f mentioned already that peraps "It doesn't fit immediately into the topic of bootctl...".

**Comment by poettering** [comment]
> Why was it added to bootctl, though? That is only available with ENABLE_BOOTLOADER==1, which is dependent on EFI and bootloader support in systemd (which we don't use). Nothing in the code for -R/-RR seems to depend on EFI and systemd bootloader integration, though.

Well, i didn't know where better to put it, and it's kinda a crucial part of the boot: the rootfs is determined then after all.

bootctl is not structly longer EFI dependent, because its our primary iface for the boot loader spec, which is not EFI specific. It does do some EFI specific stuff to though. we probably should relax the build system logic there a bit and allow it to be built even without EFI enabled. happy to take a patch...

File: src/core/main.c
Comment (poettering) [reply src/core/main.c]: should be fixed

File: man/systemd-soft-reboot.service.xml
Comment (poettering) [reply man/systemd-soft-reboot.service.xml:35]: fixed.

**Comment by poettering** [comment]
> > 
> 
> This doesn't work for sealed credentials though? These need to be un-sealable at any time when a service starts?

Yeah, but maybe we should add a different concept addressing this. i.e. seal them also to some private key stored in the tpm which we pass from one boot to the next or so. should probably discussed outside of this PR, it hink it's an somwhat independent questoin, since also needs addressing for kexec.

**Comment by bluca** [comment]
> > > 
> > 
> > 
> > This doesn't work for sealed credentials though? These need to be un-sealable at any time when a service starts?
> 
> Yeah, but maybe we should add a different concept addressing this. i.e. seal them also to some private key stored in the tpm which we pass from one boot to the next or so. should probably discussed outside of this PR, it hink it's an somwhat independent questoin, since also needs addressing for kexec.

can you add to the TODO at least so we don't forget?

**Comment by poettering** [comment]
> can you add to the TODO at least so we don't forget?

done btw

File: test/units/testsuite-82.sh
Comment (poettering) [reply test/units/testsuite-82.sh:115]: added finally.

**Comment by poettering** [comment]
This includes the fixes from #27754 btw. Once that's merged we can drop them here.

**Comment by keszybz** [comment]
> This includes the fixes from #27754 btw. Once that's merged we can drop them here.

It's been merged. Please rebase now.

**Comment by keszybz** [comment]
CI is unhappy.
jammy-ppc64el: TEST-82-SOFTREBOOT:                 FAIL     (1202 s)
In the log:
```
May 24 17:42:10 H systemd[1]: Soft-rebooting.
May 24 17:42:10 H systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interf>
May 24 17:42:10 H systemd[1]: systemd-journald.service: Releasing all stored fds
May 24 17:42:10 H systemd[1]: testsuite-82.service: Releasing all stored fds
May 24 17:42:10 H systemd[1]: Cannot stat /run/systemd/generator.late: No such file or directory
May 24 17:42:10 H systemd[1]: Bus bus-api-system: changing state RUNNING → CLOSED
May 24 17:42:10 H systemd-journald[94]: Journal stopped
```
It seems to get stuck somehow.

File: test/units/testsuite-82.sh
Comment (keszybz) [inline test/units/testsuite-82.sh]: Please no whitespace before `;`.

File: test/units/testsuite-82.sh
Comment (keszybz) [inline test/units/testsuite-82.sh]: `>"$T"`

**Review by bluca — APPROVED** [review]

**Comment by keszybz** [comment]
jammy-amd64: TEST-82-SOFTREBOOT:                 FAIL     (1202 s)
jammy-arm64: also
So this unfortunately still looks to be related.

**Comment by poettering** [comment]
woohoo seems I fixed it, setting green label again.

**Comment by bluca** [comment]
> woohoo seems I fixed it, setting green label again.

what was the issue?

**Comment by poettering** [comment]
> what was the issue?

old libmount apparently didn't carry "x-initrd.mount" over via /run/ correctly.

but i simply made /run/nextroot "extrinsic" now, much like /run/initramfs, i.e. it's now implicitly excluded from stopping at shutdown. That is what we want anyway and removes the dependency on libmount for carrying the flag over correctly.

**Comment by bluca** [comment]
> > what was the issue?
> 
> old libmount apparently didn't carry "x-initrd.mount" over via /run/ correctly.
> 
> but i simply made /run/nextroot "extrinsic" now, much like /run/initramfs, i.e. it's now implicitly excluded from stopping at shutdown. That is what we want anyway and removes the dependency on libmount for carrying the flag over correctly.

Does that work recursively? Ie if i have var mounted under nextroot will that be left alone too? The minimum libmount version needed should be documented otherwise

**Comment by poettering** [comment]
Yes, its recursive, and with this any change the libmount version isn't crucial anymore

**Pull request merged**
